src/Domain/CampaignRepository.php

```diff
@@ -443,32 +444,126 @@ public function updateCampaignFromSFData(Campaign $campaign, array $campaignData
         );
     }
 
+    /**
+     * @param array<string, string> $jsonMatchOneConditions
+     * @param array<string, string> $jsonMatchInListConditions
+     */
+    private function filterForSearch(
+        QueryBuilder $qb,
+        ?string $status,
+        array $jsonMatchOneConditions,
+        array $jsonMatchInListConditions,
+        ?string $termWildcarded
+    ): void {
+        $qb->where($qb->expr()->eq('c.hidden', '0'));
+
+        if ($status !== null) {
+            $qb->andWhere($qb->expr()->eq('c.status', ':status'));
+            $qb->setParameter('status', $status);
+        }
+
+        foreach ($jsonMatchOneConditions as $field => $value) {
+            $qb->andWhere($qb->expr()->eq("JSON_EXTRACT(c.salesforceData, '$.$field')", ':jsonMatchOne_' . $field));
+            $qb->setParameter('jsonMatchOne_' . $field, $value);
+        }
+
+        foreach ($jsonMatchInListConditions as $field => $value) {
+            $qb->andWhere($qb->expr()->like("JSON_EXTRACT(c.salesforceData, '$.$field')", ':jsonMatchInList_' . $field));
+            $qb->setParameter('jsonMatchInList_' . $field, '%' . $value . '%');
+        }
+
+        if ($termWildcarded !== null) {
+            /**
+             * @todo We'll probably want to do fulltext search and MATCH() eventually.
+            @link https://michilehr.de/full-text-search-with-mysql-and-doctrine/#3how-to-implement-a-full-text-search-with-doctrine
+             */
+            $whereSummaryMatches = "LOWER(JSON_EXTRACT(c.salesforceData, '$.summary')) LIKE LOWER(:termForWhere)";
+            $qb->andWhere(
+                $qb->expr()->orX(
+                    $qb->expr()->like('charity.name', ':termForWhere'),
+                    $qb->expr()->like('c.name', ':termForWhere'),
+                    $whereSummaryMatches,
+                )
+            );
+            $qb->setParameter('termForWhere', $termWildcarded);
+        }
+    }
+
+    private function sortForSearch(
+        QueryBuilder $qb,
+        ?string $safeSortField,
+        string $sortDirection,
+        ?string $termWildcarded
+    ): void {
+        if ($safeSortField === 'relevance') {
+            $qb->addOrderBy(
+                <<<EOT
+            CASE
+                WHEN charity.name LIKE :termForOrder THEN 20
+                WHEN c.name LIKE LOWER(:termForOrder) THEN 10
+                WHEN LOWER(JSON_EXTRACT(c.salesforceData, '$.summary')) LIKE LOWER(:termForOrder) THEN 5
+                ELSE 0
```

Comment:
might be useful to have a comment about the logic here - why the numbers 20,10,5 specifically? I guess the actual numbers don't matter as long as they're in the right order. But why does the charity name have more importance than the campaign name? Is this logic copied from the search implementation we have in SF?